File: app/src/main/java/io/trewartha/positional/position/PositionFragment.java
package io.trewartha.positional.position;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetDialog;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import io.trewartha.positional.CoordinatesFormat;
import io.trewartha.positional.Log;
import io.trewartha.positional.R;
import io.trewartha.positional.common.LocationAwareFragment;

public class PositionFragment extends LocationAwareFragment implements CompoundButton.OnCheckedChangeListener {

    private static final String TAG = PositionFragment.class.getSimpleName();

    @BindView(R.id.coordinates_view_pager) ViewPager coordinatesViewPager;
    @BindView(R.id.accuracy_value_text_view) TextView accuracyValueTextView;
    @BindView(R.id.accuracy_unit_text_view) TextView accuracyUnitTextView;
    @BindView(R.id.elevation_value_text_view) TextView elevationValueTextView;
    @BindView(R.id.elevation_unit_text_view) TextView elevationUnitTextView;
    @BindView(R.id.speed_value_text_view) TextView speedValueTextView;
    @BindView(R.id.speed_unit_text_view) TextView speedUnitTextView;
    @BindView(R.id.bearing_value_text_view) TextView bearingValueTextView;
    @BindView(R.id.bearing_unit_text_view) TextView bearingUnitTextView;

    @BindView(R.id.progress_bar) ProgressBar progressBar;
    @BindView(R.id.screen_lock_switch) ImageView screenLockSwitch;
    @BindView(R.id.copy_button) ImageView copyButton;

    private List<CoordinatesFragment> coordinatesFragments;
    private CoordinatesFormat coordinatesFormat;

    private Location location;
    private LocationFormatter locationFormatter;
    private boolean screenLock;
    private SharedPreferences sharedPreferences;
    private boolean useMetricUnits;
    private Unbinder viewUnbinder;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        coordinatesFragments = new LinkedList<>();
        locationFormatter = new LocationFormatter(context);
        sharedPreferences = context.getSharedPreferences(getString(R.string.settings_filename), Context.MODE_PRIVATE);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.position_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewUnbinder = ButterKnife.bind(this, view);

        useMetricUnits = sharedPreferences.getBoolean(getString(R.string.settings_metric_units_key), false);
        coordinatesFormat = CoordinatesFormat.valueOf(
                sharedPreferences.getString(getString(R.string.settings_coordinates_format_key), CoordinatesFormat.DMS.name())
        );
        screenLock = sharedPreferences.getBoolean(getString(R.string.settings_screen_lock_key), false);
        screenLockSwitch.setSelected(screenLock);

        final CoordinatesFragmentPagerAdapter coordinatesPagerAdapter = new CoordinatesFragmentPagerAdapter(
                getChildFragmentManager()
        );
        coordinatesViewPager.setAdapter(coordinatesPagerAdapter);
        coordinatesViewPager.setOffscreenPageLimit(coordinatesPagerAdapter.getCount());
        coordinatesViewPager.setCurrentItem(getCoordinatesFragmentIndex(coordinatesFormat), true);
        coordinatesViewPager.addOnPageChangeListener(new CoordinatesPageChangeListener());

        updateLocationViews(null);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        viewUnbinder.unbind();
    }

    @Override
    public void onAttachFragment(Fragment fragment) {
        if (fragment instanceof CoordinatesFragment) {
            coordinatesFragments.add((CoordinatesFragment) fragment);
        }
    }

    @Override
    public void onLocationChanged(@Nullable Location location) {
        this.location = location;
        if (location == null) {
            progressBar.setVisibility(View.VISIBLE);
        } else {
            logLocation(location);
            if (progressBar.getVisibility() == View.VISIBLE) {
                progressBar.setVisibility(View.INVISIBLE);
            }
            updateLocationViews(location);
        }
    }

    @OnClick(R.id.copy_button)
    public void onCopyClicked() {
        final BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(getContext());
        bottomSheetDialog.setCanceledOnTouchOutside(true);
        bottomSheetDialog.setCancelable(true);
        bottomSheetDialog.setTitle(R.string.settings_copy_coordinates_title);
        bottomSheetDialog.setContentView(R.layout.coordinates_copy_fragment);

        CoordinatesCopier coordinatesCopier = new CoordinatesCopier(new OnCoordinatesCopiedListener() {
            @Override
            public void onCopy() {
                bottomSheetDialog.dismiss();
            }
        });

        View bothTextView = bottomSheetDialog.findViewById(R.id.coordinates_copy_both_text_view);
        View latitudeTextView = bottomSheetDialog.findViewById(R.id.coordinates_copy_latitude_text_view);
        View longitudeTextView = bottomSheetDialog.findViewById(R.id.coordinates_copy_longitude_text_view);

        if (bothTextView == null || latitudeTextView == null || longitudeTextView == null) {
            Toast.makeText(getContext(), R.string.copied_coordinates_failure, Toast.LENGTH_SHORT).show();
            return;
        }

        bothTextView.setOnClickListener(coordinatesCopier);
        latitudeTextView.setOnClickListener(coordinatesCopier);
        longitudeTextView.setOnClickListener(coordinatesCopier);

        bottomSheetDialog.show();
    }

    @OnClick({R.id.elevation_unit_text_view, R.id.speed_unit_text_view, R.id.accuracy_unit_text_view})
    public void onDistanceUnitClicked() {
        useMetricUnits = !useMetricUnits;
        updateLocationViews(location);
        setBooleanPreference(getString(R.string.settings_metric_units_key), useMetricUnits);
    }

    @OnClick(R.id.screen_lock_switch)
    public void onScreenLockClicked() {
        screenLock = !screenLock;
        screenLockSwitch.setSelected(screenLock);
        setBooleanPreference(getString(R.string.settings_screen_lock_key), screenLock);
        final int textRes = screenLock ? R.string.screen_lock_on : R.string.screen_lock_off;
        Toast.makeText(getContext(), textRes, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onCheckedChanged(CompoundButton compoundButton, boolean checked) {
        if (compoundButton.getId() == R.id.screen_lock_switch) {
            setBooleanPreference(getString(R.string.settings_screen_lock_key), checked);
            lockScreen(checked);
        }
    }

    private int getCoordinatesFragmentIndex(@NonNull CoordinatesFormat coordinatesFormat) {
        switch (coordinatesFormat) {
            case DECIMAL:
                return 0;
            case DMS:
                return 1;
            case UTM:
                return 2;
            case MGRS:
                return 3;
            default:
                return 0;
        }
    }

    private void lockScreen(boolean lock) {
        if (lock) {
            getActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        } else {
            getActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    private void logLocation(@NonNull Location location) {
        Log.debug(TAG, "Location received:");
        Log.debug(TAG, "       Latitude: " + location.getLatitude());
        Log.debug(TAG, "      Longitude: " + location.getLongitude());
        Log.debug(TAG, "       Accuracy: " + (location.hasAccuracy() ? location.getAccuracy() + " m" : "none"));
        Log.debug(TAG, "        Bearing: " + (location.hasBearing() ? location.getBearing() + "°" : "none"));
        Log.debug(TAG, "      Elevation: " + (location.hasAltitude() ? location.getAltitude() + " m/s" : "none"));
        Log.debug(TAG, "          Speed: " + (location.hasSpeed() ? location.getSpeed() + " m/s" : "none"));
    }

    private void updateLocationViews(@Nullable Location location) {
        updateCoordinatesFragments(location);
        accuracyValueTextView.setText(locationFormatter.getAccuracy(location, useMetricUnits));
        accuracyUnitTextView.setText(locationFormatter.getDistanceUnit(useMetricUnits));
        bearingValueTextView.setText(locationFormatter.getBearing(location));
        elevationValueTextView.setText(locationFormatter.getElevation(location, useMetricUnits));
        elevationUnitTextView.setText(locationFormatter.getDistanceUnit(useMetricUnits));
        speedValueTextView.setText(locationFormatter.getSpeed(location, useMetricUnits));
        speedUnitTextView.setText(locationFormatter.getSpeedUnit(useMetricUnits));
    }

    private void setBooleanPreference(@NonNull String key, boolean value) {
        Log.info(TAG, "Saving " + key + " preference as " + value);
        sharedPreferences.edit().putBoolean(key, value).apply();
    }

    private void setStringPreference(@NonNull String key, @Nullable String value) {
        Log.info(TAG, "Saving " + key + " preference as " + value);
        sharedPreferences.edit().putString(key, value).apply();
    }

    private void updateCoordinatesFragments(@Nullable Location location) {
        final double latitude, longitude;
        if (location == null) {
            latitude = 0.0;
            longitude = 0.0;
        } else {
            latitude = location.getLatitude();
            longitude = location.getLongitude();
        }
        for (CoordinatesFragment coordinatesFragment : coordinatesFragments) {
            coordinatesFragment.setCoordinates(latitude, longitude);
        }
    }

    private class CoordinatesPageChangeListener implements ViewPager.OnPageChangeListener {

        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            // Don't do anything here
        }

        @Override
        public void onPageSelected(int position) {
            switch (position) {
                case 0:
                    coordinatesFormat = CoordinatesFormat.DECIMAL;
                    break;
                case 1:
                    coordinatesFormat = CoordinatesFormat.DMS;
                    break;
                case 2:
                    coordinatesFormat = CoordinatesFormat.UTM;
                    break;
                case 3:
                    coordinatesFormat = CoordinatesFormat.MGRS;
                    break;
                default:
                    coordinatesFormat = CoordinatesFormat.DECIMAL;
                    break;
            }
            setStringPreference(getString(R.string.settings_coordinates_format_key), coordinatesFormat.name());
        }

        @Override
        public void onPageScrollStateChanged(int state) {
            // Don't do anything here
        }
    }

    private class CoordinatesCopier implements View.OnClickListener {

        private OnCoordinatesCopiedListener onCoordinatesCopiedListener;

        CoordinatesCopier(@Nullable OnCoordinatesCopiedListener onCoordinatesCopiedListener) {
            this.onCoordinatesCopiedListener = onCoordinatesCopiedListener;
        }

        @Override
        public void onClick(View v) {
            if (location == null) {
                Toast.makeText(getContext(), R.string.copied_coordinates_failure, Toast.LENGTH_SHORT).show();
                return;
            }

            ClipboardManager clipboardManager = (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
            String clipDataLabel = getString(R.string.copied_coordinates_label);
            String clipDataText = "";
            String toastText = "";

            switch (v.getId()) {
                case R.id.coordinates_copy_both_text_view:
                    clipDataText = String.format(Locale.US, "%f, %f", location.getLatitude(), location.getLongitude());
                    toastText = getString(R.string.copied_coordinates_both_success);
                    break;
                case R.id.coordinates_copy_latitude_text_view:
                    clipDataText = String.format(Locale.US, "%f", location.getLatitude());
                    toastText = getString(R.string.copied_coordinates_latitude_success);
                    break;
                case R.id.coordinates_copy_longitude_text_view:
                    clipDataText = String.format(Locale.US, "%f", location.getLongitude());
                    toastText = getString(R.string.copied_coordinates_longitude_success);
                    break;
            }

            ClipData coordinatesClipData = ClipData.newPlainText(clipDataLabel, clipDataText);
            clipboardManager.setPrimaryClip(coordinatesClipData);

            Toast.makeText(getContext(), toastText, Toast.LENGTH_SHORT).show();

            if (onCoordinatesCopiedListener != null) {
                onCoordinatesCopiedListener.onCopy();
            }
        }
    }

    private interface OnCoordinatesCopiedListener {
        void onCopy();
    }
}

File: app/src/main/java/io/trewartha/positional/ui/PositionFragment.kt
package io.trewartha.positional.ui

import android.content.*
import android.location.Location
import android.os.Bundle
import android.support.design.widget.BottomSheetDialog
import android.support.v4.app.Fragment
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.CompoundButton
import android.widget.Toast
import com.google.android.gms.location.LocationRequest
import io.trewartha.positional.location.CoordinatesFormat
import io.trewartha.positional.Log
import io.trewartha.positional.R
import io.trewartha.positional.location.LocationFormatter
import kotlinx.android.synthetic.main.position_fragment.*
import java.util.*


class PositionFragment : LocationAwareFragment(), CompoundButton.OnCheckedChangeListener {

    companion object {
        private val TAG = "PositionFragment"
        private val LOCATION_UPDATE_INTERVAL = 5000L // ms
        private val LOCATION_UPDATE_MAX_WAIT_TIME = 10000L // ms
        private val LOCATION_UPDATE_PRIORITY = LocationRequest.PRIORITY_HIGH_ACCURACY
    }

    private var coordinatesFragments: MutableList<CoordinatesFragment> = LinkedList()

    private lateinit var coordinatesFormat: CoordinatesFormat
    private lateinit var locationFormatter: LocationFormatter
    private lateinit var sharedPreferences: SharedPreferences

    private var location: Location? = null
    private var screenLock: Boolean = false
    private var useMetricUnits: Boolean = false

    override fun onAttach(context: Context) {
        super.onAttach(context)
        locationFormatter = LocationFormatter(context)
        sharedPreferences = context.getSharedPreferences(getString(R.string.settings_filename), Context.MODE_PRIVATE)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.position_fragment, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        copyButton.setOnClickListener { onCopyClicked() }
        shareButton.setOnClickListener { onShareClicked() }

        val unitOnClickListener = View.OnClickListener { onDistanceUnitClicked() }
        elevationUnitTextView.setOnClickListener(unitOnClickListener)
        speedUnitTextView.setOnClickListener(unitOnClickListener)
        accuracyUnitTextView.setOnClickListener(unitOnClickListener)

        screenLockSwitch.setOnClickListener { onScreenLockClicked() }

        useMetricUnits = sharedPreferences.getBoolean(getString(R.string.settings_metric_units_key), false)
        coordinatesFormat = CoordinatesFormat.valueOf(
                sharedPreferences.getString(getString(R.string.settings_coordinates_format_key), CoordinatesFormat.DMS.name)
        )
        screenLock = sharedPreferences.getBoolean(getString(R.string.settings_screen_lock_key), false)
        screenLockSwitch.isSelected = screenLock

        val coordinatesPagerAdapter = CoordinatesFragmentPagerAdapter(
                childFragmentManager
        )
        coordinatesViewPager.adapter = coordinatesPagerAdapter
        coordinatesViewPager.offscreenPageLimit = coordinatesPagerAdapter.count
        coordinatesViewPager.setCurrentItem(getCoordinatesFragmentIndex(coordinatesFormat), true)
        coordinatesViewPager.addOnPageChangeListener(CoordinatesPageChangeListener())

        updateLocationViews(null)
    }

    override fun onAttachFragment(fragment: Fragment?) {
        if (fragment is CoordinatesFragment) {
            coordinatesFragments.add(fragment)
        }
    }

    override fun onLocationChanged(location: Location?) {
        this.location = location
        if (location == null) {
            progressBar.visibility = View.VISIBLE
        } else {
            if (progressBar.visibility == View.VISIBLE) {
                progressBar.visibility = View.INVISIBLE
            }
            updateLocationViews(location)
        }
    }

    override fun getLocationUpdateInterval(): Long {
        return LOCATION_UPDATE_INTERVAL
    }

    override fun getLocationUpdateMaxWaitTime(): Long {
        return LOCATION_UPDATE_MAX_WAIT_TIME
    }

    override fun getLocationUpdatePriority(): Int {
        return LOCATION_UPDATE_PRIORITY
    }

    override fun onCheckedChanged(compoundButton: CompoundButton, checked: Boolean) {
        if (compoundButton.id == R.id.screenLockSwitch) {
            setBooleanPreference(getString(R.string.settings_screen_lock_key), checked)
            lockScreen(checked)
        }
    }

    private fun getCoordinatesFragmentIndex(coordinatesFormat: CoordinatesFormat): Int {
        return when (coordinatesFormat) {
            CoordinatesFormat.DECIMAL -> 0
            CoordinatesFormat.DMS -> 1
            CoordinatesFormat.UTM -> 2
            CoordinatesFormat.MGRS -> 3
        }
    }

    private fun lockScreen(lock: Boolean) {
        if (lock) {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    private fun onCopyClicked() {
        val safeLocation = location
        if (safeLocation == null) {
            Toast.makeText(context, R.string.copied_coordinates_failure, Toast.LENGTH_SHORT).show()
            return
        }

        val bottomSheetDialog = BottomSheetDialog(context)
        bottomSheetDialog.setCanceledOnTouchOutside(true)
        bottomSheetDialog.setCancelable(true)
        bottomSheetDialog.setTitle(R.string.settings_copy_coordinates_title)
        bottomSheetDialog.setContentView(R.layout.coordinates_copy_fragment)

        val coordinatesCopier = CoordinatesCopier(context, safeLocation, object : OnCoordinatesCopiedListener {
            override fun onCopy() {
                bottomSheetDialog.dismiss()
            }
        })

        val bothTextView = bottomSheetDialog.findViewById<View>(R.id.coordinatesCopyBothTextView)
        val latitudeTextView = bottomSheetDialog.findViewById<View>(R.id.coordinatesCopyLatitudeTextView)
        val longitudeTextView = bottomSheetDialog.findViewById<View>(R.id.coordinatesCopyLongitudeTextView)

        if (bothTextView == null || latitudeTextView == null || longitudeTextView == null) {
            Toast.makeText(context, R.string.copied_coordinates_failure, Toast.LENGTH_SHORT).show()
            return
        }

        bothTextView.setOnClickListener(coordinatesCopier)
        latitudeTextView.setOnClickListener(coordinatesCopier)
        longitudeTextView.setOnClickListener(coordinatesCopier)

        bottomSheetDialog.show()
    }

    private fun onDistanceUnitClicked() {
        useMetricUnits = !useMetricUnits
        updateLocationViews(location)
        setBooleanPreference(getString(R.string.settings_metric_units_key), useMetricUnits)
    }

    private fun onScreenLockClicked() {
        screenLock = !screenLock
        screenLockSwitch.isSelected = screenLock
        setBooleanPreference(getString(R.string.settings_screen_lock_key), screenLock)
        val textRes = if (screenLock) R.string.screen_lock_on else R.string.screen_lock_off
        Toast.makeText(context, textRes, Toast.LENGTH_SHORT).show()
    }

    private fun onShareClicked() {
        val safeLocation = location
        if (safeLocation == null) {
            Toast.makeText(context, R.string.shared_coordinates_failure, Toast.LENGTH_SHORT).show()
            return
        }

        val coordinatesText = locationFormatter.getCoordinates(
                safeLocation.latitude,
                safeLocation.longitude,
                coordinatesFormat
        )

        val sendIntent = Intent(Intent.ACTION_SEND)
        sendIntent.putExtra(Intent.EXTRA_TEXT, coordinatesText)
        sendIntent.type = "text/plain"
        startActivity(sendIntent)
    }

    private fun updateLocationViews(location: Location?) {
        updateCoordinatesFragments(location)
        accuracyValueTextView.text = locationFormatter.getAccuracy(location, useMetricUnits)
        accuracyUnitTextView.text = locationFormatter.getDistanceUnit(useMetricUnits)
        bearingValueTextView.text = locationFormatter.getBearing(location)
        elevationValueTextView.text = locationFormatter.getElevation(location, useMetricUnits)
        elevationUnitTextView.text = locationFormatter.getDistanceUnit(useMetricUnits)
        speedValueTextView.text = locationFormatter.getSpeed(location, useMetricUnits)
        speedUnitTextView.text = locationFormatter.getSpeedUnit(useMetricUnits)
    }

    private fun setBooleanPreference(key: String, value: Boolean) {
        Log.info(TAG, "Saving $key preference as $value")
        sharedPreferences.edit().putBoolean(key, value).apply()
    }

    private fun setStringPreference(key: String, value: String?) {
        Log.info(TAG, "Saving $key preference as $value")
        sharedPreferences.edit().putString(key, value).apply()
    }

    private fun updateCoordinatesFragments(location: Location?) {
        val latitude: Double
        val longitude: Double
        if (location == null) {
            latitude = 0.0
            longitude = 0.0
        } else {
            latitude = location.latitude
            longitude = location.longitude
        }
        coordinatesFragments.forEach {
            it.setCoordinates(latitude, longitude)
        }
    }

    private inner class CoordinatesPageChangeListener : ViewPager.OnPageChangeListener {

        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            // Don't do anything here
        }

        override fun onPageSelected(position: Int) {
            coordinatesFormat = when (position) {
                0 -> CoordinatesFormat.DECIMAL
                1 -> CoordinatesFormat.DMS
                2 -> CoordinatesFormat.UTM
                3 -> CoordinatesFormat.MGRS
                else -> CoordinatesFormat.DECIMAL
            }
            setStringPreference(getString(R.string.settings_coordinates_format_key), coordinatesFormat.name)
        }

        override fun onPageScrollStateChanged(state: Int) {
            // Don't do anything here
        }
    }

    private class CoordinatesCopier(val context: Context, val location: Location, val onCoordinatesCopiedListener: OnCoordinatesCopiedListener?) : View.OnClickListener {

        override fun onClick(v: View) {
            val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clipDataLabel = context.getString(R.string.copied_coordinates_label)
            var clipDataText = ""
            var toastText = ""

            when (v.id) {
                R.id.coordinatesCopyBothTextView -> {
                    clipDataText = String.format(Locale.US, "%f, %f", location.latitude, location.longitude)
                    toastText = context.getString(R.string.copied_coordinates_both_success)
                }
                R.id.coordinatesCopyLatitudeTextView -> {
                    clipDataText = String.format(Locale.US, "%f", location.latitude)
                    toastText = context.getString(R.string.copied_coordinates_latitude_success)
                }
                R.id.coordinatesCopyLongitudeTextView -> {
                    clipDataText = String.format(Locale.US, "%f", location.longitude)
                    toastText = context.getString(R.string.copied_coordinates_longitude_success)
                }
            }

            clipboardManager.primaryClip = ClipData.newPlainText(clipDataLabel, clipDataText)

            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show()

            onCoordinatesCopiedListener?.onCopy()
        }
    }

    private interface OnCoordinatesCopiedListener {
        fun onCopy()
    }
}
